File: week3/java_week3/Date/Date.java
enum Months{
    JAN(1), FEB(2), MAR(3), 
    APR(4), MAY(5), JUN(6), 
    JUL(7), AUG(8), SEP(9), 
    OCT(10), NOV(11), DEC(12);

    private int month;

    Months(int month) {
        this.month = month;
    }

    public int getValue() {
        return this.month;
    }
}

public class Date {
    private int day;
    private int month;
    private int year;
    public String text;
    
    private void validate(int day, int month, int year) throws InvalidDateException {
        boolean leap;

        if ( year <= 0 ) {
            throw new InvalidDateException("Invalid year.");
        }

        leap = ( (year % 4 == 0 && year % 100 != 0) || year % 400 == 0);

        if ( month < Months.JAN.getValue() || month > Months.DEC.getValue() ) {
            throw new InvalidDateException("Invalid month.");
        }
        
        if ( month == Months.APR.getValue() || month == Months.JUN.getValue() || month == Months.SEP.getValue() || month == Months.NOV.getValue() ) {
            if ( day < 1 || day > 30 ) {
                throw new InvalidDateException("Invalid day of month.");
            }
        } else if ( month == Months.FEB.getValue() && leap ) {
            if ( day < 1 || day > 29 ) {
                throw new InvalidDateException("Invalid day in Feb (leap year).");
            }
        } else if ( month == Months.FEB.getValue() && !leap ) {
            if ( day < 1 || day > 28 ) {
                throw new InvalidDateException("Invalid day of month.");
            }
        } else {
            if ( day < 1 || day > 31 ) {
                throw new InvalidDateException("Invalid day of month.");
            }
        }
    }

    public Date(int day, int month, int year) throws InvalidDateException {
        this.day = day;
        this.month = month;
        this.year = year;
        this.validate(day, month, year);
    }

    public Date() throws InvalidDateException {
        this(1, 1, 1970);
    }

    public int getDay() {
        return this.day;
    }

    public int getMonth() {
        return this.month;
    }

    public int getYear() {
        return this.year;
    }

    public String getDate() {
        StringBuilder date = new StringBuilder();

        if ( this.day < 10 ) {
            date.append('0').append(this.day).append('-');
        } else {
            date.append(this.day).append('-');
        }

        if ( this.month < 10 ) {
            date.append('0').append(this.month).append('-');
        } else {
            date.append(this.month).append('-');
        }

        date.append(this.year);

        return date.toString();
    }
}

File: week3/cpp_hw_week3/Date/Date.cpp
#include "Date.h"

enum MONTHS {
    JAN = 1,
    FEB,
    MAR,
    APR,
    MAY,
    JUN,
    JUL,
    AUG,
    SEP,
    OCT,
    NOV,
    DEC
};

void Date::validate(int day, int month, int year) {
    bool leap;

    if ( year <= 0 ) {
        throw InvalidDate("Invalid year.");
    }

    leap = ( (year % 4 == 0 && year % 100 != 0) || year % 400 == 0);

    if ( month < JAN || month > DEC ) {
        throw InvalidDate("Month should be in range 1 - 12.");
    }

    if ( month == APR || month == JUN || month == SEP || month == NOV ) {
        if ( day < 1 || day > 30 ) {
            throw InvalidDate("Day should be in range 1 - 30.");
        }
    } else if ( month == FEB && leap ) {
        if ( day < 1 || day > 29 ) {
            throw InvalidDate("Day should be in range 1 - 29.");
        }
    } else if ( month == FEB && !leap ) {
        if ( day < 1 || day > 28 ) {
            throw InvalidDate("Day should be in range 1 - 28.");
        }
    } else {
        if ( day < 1 || day > 31 ) {
            throw InvalidDate("Day should be in range 1 - 31.");
        }
    }
}

Date::Date(int day, int month, int year) {
    this->day = day;
    this->month = month;
    this->year = year;
    this->validate(day, month, year);
}

Date::~Date() {}

int Date::getDay() const {
    return this->day;
}

int Date::getMonth() const {
    return this->month;
}

int Date::getYear() const {
    return this->year;
}

std::ostream& operator<<(std::ostream& out, const Date& date) {
    out << date.getDay() << "-" << date.getMonth() << "-" << date.getYear();
    return out;
}